hpx/lcos/detail/future_data.hpp

```diff
@@ -300,8 +301,11 @@ namespace detail
             wait(ec);
             if (ec) return NULL;
 
+            boost::unique_lock<mutex_type> l(mtx_);
```

Comment:
This lock is unnecessary. Once a future has been made ready, which is a postcondition of `wait`, either:

- there is only one writer (`future`), or
- there are multiple readers only (`shared_future`, lock hurts concurrency)